File: Assets/RFramework/Game/LuaScript/Game/Module/ModuleManager.lua
-- 模组管理器，代码思路参考ModuleManaget.cs 

local ModuleManager_Class = BaseClass()

function ModuleManager_Class:ctor()
	self.m_allModules = {}
	self.m_updateModules = {}
	self.m_lateUpdateModules = {}
	self.m_startQueue = {}
	self.m_starting = nil
end

function ModuleManager_Class:AddModule(module)
	if module == nil then return end
	table.insert(self.m_startQueue, module)
	table.insert(self.m_allModules, module)
end

function ModuleManager_Class:CheckStartComplete()
	if self.m_starting == nil and #self.m_startQueue == 0 then
		return
	end

	if self.m_starting == nil then
		self.m_starting = table.remove(self.m_startQueue, 1)
		self.m_starting:OnStart()
	end
	if self.m_starting.IsComplete then
		self.m_starting:OnFinishStart()
		if self.m_starting.IsUpdate then
			table.insert(self.m_updateModules, self.m_starting)
		end
		if self.m_starting.OnLateUpdate then
			table.insert(self.m_lateUpdateModules, self.m_starting)
		end

		if #self.m_startQueue > 0 then
			self.m_starting = table.remove(self.m_startQueue, 1)
			self.m_starting:OnStart()
		else
			self.m_starting = nil
		end
	elseif self.m_starting.OnUpdate then
		self.m_starting:OnUpdate()
	end
end

function ModuleManager_Class:UpdateModules()
	self:CheckStartComplete()
	for _, v in self.m_updateModules do
		v:OnUpdate()
	end
end

function ModuleManager_Class:LateUpdateModules()
	for _, v in self.m_lateUpdateModules do
		v:OnLateUpdate()
	end
end

function ModuleManager_Class:DestroyModules()
	for _, v in self.m_allModules do
		v:OnDestroy()
	end
end

ModuleManager = ModuleManager_Class.New()

File: Assets/RFramework/RFramework/Module/ModuleManager.cs
using System.Collections;
using System.Collections.Generic;
using RFramework;
using UnityEngine;

namespace RFramework
{
    /// <summary>
    /// 模组管理器，负责所有模组的初始化、Update、销毁
    /// </summary>
    public class ModuleManager
    {
        private List<BaseModule> m_allModules; // 所有Module
        private List<BaseModule> m_updateModules; // 需要Update的Module
        private List<LateUpdateModuleInterface> m_lateUpdateModules; // 需要LateUpdate的Module
        private Queue<BaseModule> m_startQueue; // 初始化队列
        private BaseModule m_starting; // 当前正在初始化的Module

        public ModuleManager()
        {
            m_allModules = new List<BaseModule>();
            m_updateModules = new List<BaseModule>();
            m_lateUpdateModules = new List<LateUpdateModuleInterface>();
            m_startQueue = new Queue<BaseModule>();
        }

        public void AddModule(BaseModule module)
        {
            if (m_allModules.Contains(module)) return;
            m_allModules.Add(module);
            m_startQueue.Enqueue(module);
        }

        /// <summary>
        /// 检查当前starting是否已完成，根据IsComplete判断
        /// </summary>
        public void CheckStartComplete()
        {
            if (m_starting == null && m_startQueue.Count == 0) return;

            if (m_starting == null)
            {
                m_starting = m_startQueue.Dequeue();
                m_starting.OnStart();
            }

            if (m_starting.IsComplete)
            {
                // 初始化完成
                m_starting.OnFinishStart();
                if (m_starting.IsUpdate)
                {
                    m_updateModules.Add(m_starting);
                }

                var lateUpdate = m_starting as LateUpdateModuleInterface;
                if (lateUpdate != null)
                {
                    m_lateUpdateModules.Add(lateUpdate);
                }

                // 下一个
                if (m_startQueue.Count > 0)
                {
                    m_starting = m_startQueue.Dequeue();
                    m_starting.OnStart();
                }
                else
                {
                    m_starting = null;
                }
            }
            else // if(m_starting.IsUpdate)
            {
                m_starting.OnUpdate();
            }
        }

        /// <summary>
        /// 初始化以及OnUpdate
        /// </summary>
        public void UpdateModules()
        {
            CheckStartComplete();
            foreach (var module in m_updateModules)
            {
                module.OnUpdate();
            }
        }

        /// <summary>
        /// OnLateUpdate
        /// </summary>
        public void LateUpdateModules()
        {
            foreach (var module in m_lateUpdateModules)
            {
                module.OnLateUpdate();
            }
        }

        /// <summary>
        /// OnDestroy
        /// </summary>
        public void DestroyModules()
        {
            foreach (var module in m_allModules)
            {
                module.OnDestroy();
            }
        }
    }
}
